### PR TITLE
[flang][OpenMP][OMPIRBuilder][mlir] Optionally pass reduction vars by ref

### PR DESCRIPTION
**Pull request by tblah**
Previously reduction variables were always passed by value into and out of the initialization and combiner regions of the OpenMP reduction declare operation.

This worked well for reductions of primitive types (and might perform better than passing by reference). But passing by reference will be useful for array and derived type reductions (e.g. to move allocation inside of the init region).

Passing reductions by reference requires different LLVM-IR generation when lowering from MLIR because some of the loads/stores/allocations will now be moved inside of the init and combiner regions. This alternate code generation is requested using a new attribute to omp.wsloop and omp.parallel.

Existing lowerings from mlir are unaffected (these will continue to use the by-value argument passing.

Flang will continue to pass by-value argument passing for trivial types unless a (hidden) command line argument is supplied. Non-trivial types will always use the by-ref lowering.

Array reductions are not ready yet (but are coming very soon). In the meantime, this is tested by forcing existing reductions to use by-ref.

Commit series for by-ref OpenMP reductions 3/3

### CODE REVIEW

**Comment by tblah** [comment]
co-authored with @Leporacanthicus  (github seems to have taken the tag out of the commit message but shows it in the UI)

**Review by Leporacanthicus — APPROVED** [review]
LGTM.

File: flang/lib/Lower/OpenMP/ReductionProcessor.cpp [hunk 1]
```diff
@@ -294,14 +322,41 @@ mlir::omp::ReductionDeclareOp ReductionProcessor::createReductionDecl(
...
+  if (forceByrefReduction)
+    return true;
+
+  for (mlir::Value reductionVar : reductionVars) {
+    if (auto declare =
+            mlir::dyn_cast<hlfir::DeclareOp>(reductionVar.getDefiningOp()))
+      reductionVar = declare.getMemref();
+
+    if (!fir::isa_trivial(fir::unwrapRefType(reductionVar.getType())))
+      return true;
```
Comment (DavidTruby) [inline flang/lib/Lower/OpenMP/ReductionProcessor.cpp:355]: Does this imply that all reductions on a clause have to be by ref or by val? E.g. if we have an array reduction on the clause does that mean an integer reduction also changes to byref?

Comment (tblah) [reply flang/lib/Lower/OpenMP/ReductionProcessor.cpp:355]: Yes it does. I did this to keep things simpler. Currently byref vs byval is toggled over the whole wsloop or parallel region. A more sophisticated implementation could instead track this per reduction argument. I chose not to do this to keep things simple.

I suspect that in most cases, if an integer reduction and an array reduction are used together, the array reduction would take long enough that the performance loss from doing the integer reduction by reference would not be significant. I have not measured this.

Comment (DavidTruby) [reply flang/lib/Lower/OpenMP/ReductionProcessor.cpp:355]: That's probably true wrt performance, but I believe when it comes to `openmp target` reductions doing reductions on basic types will affect correctness. As far as I remember you do not need to ensure manually that for example the INTEGER exists on the target device, whereas you do for an array (and would need to if the integer is passed by reference).

I think fixing that in a subsequent patch is probably fine though, as long as we add a TODO mentioning that ideally it should be considered separately per argument.

**Review by DavidTruby — APPROVED** [review]
LGTM

**Pull request merged**